File: src/ObjectManager/Object.Tes/FilePacks/Records/345-CONT.Container.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class CONTRecord : Record, IHaveEDID, IHaveMODL
    {
        // TESX
        public class DATAField
        {
            public byte Flags; // flags 0x0001 = Organic, 0x0002 = Respawns, organic only, 0x0008 = Default, unknown
            public float Weight;

            public DATAField(UnityBinaryReader r, uint dataSize, GameFormatId formatId)
            {
                if (formatId == GameFormatId.Tes3)
                {
                    Weight = r.ReadLESingle();
                    return;
                }
                Flags = r.ReadByte();
                Weight = r.ReadLESingle();
            }

            public void FLAGField(UnityBinaryReader r, uint dataSize)
            {
                Flags = (byte)r.ReadLEUInt32();
            }
        }

        public override string ToString() => $"CONT: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model
        public STRVField FULL; // Container Name
        public DATAField DATA; // Container Data
        public FMIDField<SCPTRecord>? SCRI;
        public List<CNTOField> CNTOs = new List<CNTOField>();
        // TES4
        public FMIDField<SOUNRecord> SNAM; // Open sound
        public FMIDField<SOUNRecord> QNAM; // Close sound

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, uint dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL":
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "DATA":
                case "CNDT": DATA = new DATAField(r, dataSize, formatId); return true;
                case "FLAG": DATA.FLAGField(r, dataSize); return true;
                case "CNTO":
                case "NPCO": CNTOs.Add(new CNTOField(r, dataSize, formatId)); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "SNAM": SNAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                case "QNAM": QNAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-CONT.Container.swift
//
//  CONTRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class CONTRecord: Record, IHaveEDID, IHaveMODL {
    // TESX
    public class DATAField {
        public var flags: UInt8 // flags 0x0001 = Organic, 0x0002 = Respawns, organic only, 0x0008 = Default, unknown
        public let weight: Float

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                flags = 0
                weight = r.readLESingle()
                return
            }
            flags = r.readByte()
            weight = r.readLESingle()
        }

        func FLAGField(_ r: BinaryReader, _ dataSize: Int) {
            flags = UInt8(r.readLEUInt32())
        }
    }

    public override var description: String { return "CONT: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model
    public var FULL: STRVField! // Container Name
    public var DATA: DATAField! // Container Data
    public var SCRI: FMIDField<SCPTRecord>? = nil
    public var CNTOs = [CNTOField]()
    // TES4
    public var SNAM: FMIDField<SOUNRecord>! // Open sound
    public var QNAM: FMIDField<SOUNRecord>! // Close sound
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL",
             "FNAM": FULL = r.readSTRV(dataSize)
        case "DATA",
             "CNDT": DATA = DATAField(r, dataSize, format)
        case "FLAG": DATA.FLAGField(r, dataSize)
        case "CNTO",
             "NPCO": CNTOs.append(CNTOField(r, dataSize, for: format))
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "SNAM": SNAM = FMIDField<SOUNRecord>(r, dataSize)
        case "QNAM": QNAM = FMIDField<SOUNRecord>(r, dataSize)
        default: return false
        }
        return true
    }
}
